### ISSUE TITLE
Bad vertex lighting with station15 tree

### ISSUE THREAD

**Issue by slipher** [Unvanquished/Unvanquished#3179]
Not [that tree](https://unvanquished.net/story-of-a-tree/), but the evergreen one:

![unvanquished_2024-11-03_170210_000](https://github.com/user-attachments/assets/ad769321-a38a-4752-b802-8b2df85a3974)

Some branches are inexplicably much more brightly lit than the others. It has looked that way since the tree was introduced in Unvanquished 0.29/station15 1.0.

The first version of the shader looked like this (commented lines elided):
```
textures/station15_custom/tree1_leaves
{
	qer_editorImage textures/station15_custom_src/tree1_d
	qer_trans           0.50
	surfaceparm alphashadow
	surfaceparm trans
	surfaceparm nonsolid
	surfaceparm nomarks
	surfaceparm nolightmap
	cull                none
	{
		map       textures/station15_custom_src/tree1_d
		rgbGen vertex
		alphaFunc GE128
	}
}
```

Now it's this.

```
models/station15/mundo_tree2_leaves
{
	qer_editorImage models/station15/mundo_tree2_d
	qer_alphaFunc greater .5

	surfaceparm alphashadow
	surfaceparm trans
	surfaceparm nonsolid
	surfaceparm nomarks
	surfaceparm nolightmap
	cull none

	{
		diffuseMap models/station15/mundo_tree2_d
		normalMap models/station15/mundo_tree2_n
		specularMap models/station15/mundo_tree2_s
		rgbGen vertex
		alphaFunc GE128
	}
}
```

Is vertex lighting typically broken for irregularly-shaped models? Is there some reason that vertex lighting needs to be used at all?

**Comment by VReaperV** [comment]
Might be a duplicate of https://github.com/DaemonEngine/Daemon/issues/119, hard to tell if it's thesame issue from screenshots.

**Comment by slipher** [comment]
> Might be a duplicate of [DaemonEngine/Daemon#119](https://github.com/DaemonEngine/Daemon/issues/119), hard to tell if it's the same issue from screenshots.

Nope. #119 only affects one spot at the bottom of the tree. That issue is easy to see in-game with `r_clear` disabled, but hard to understand from screenshots. This issue is easy to see in a screenshot: I'm just talking about how some branches are much lighter than others.

**Comment by slipher** [comment]
Showing the precomputed lighting colors with `r_showVertexColors`:
![unvanquished_2024-11-03_230410_000](https://github.com/user-attachments/assets/0bb045ed-6c59-42d7-b5c2-f82ff981e443)

The other instance of this tree model in the same room has a similar pattern.

**Comment by VReaperV** [comment]
Depth buffer after depth pass on the tree:
![station15tree_depth](https://github.com/user-attachments/assets/03a70b85-bf4b-4c1b-aee5-5b86dc81508e)

A lot of the vertexes have very low colour values, <= 10. Some others have colour values up to about 60, which is probably where the discrepancy comes from, since they're just multiplied by the diffuse.

Here's the colour buffer after rendering this tree, with colour values normalized in range of [0.0, 0.00390625]:
![station15tree_colour](https://github.com/user-attachments/assets/f970ed51-344e-44d1-bee1-cac134239bf4)

**Comment by VReaperV** [comment]
(comment copied from https://github.com/DaemonEngine/Daemon/issues/119)

**Comment by VReaperV** [comment]
I believe this is therefore an asset bug.

**Comment by slipher** [comment]
Yes, what we want to do is somehow make q3map2 stop outputting garbage light values in the vertex colors. Or alternatively try it with lightmapping instead (the other tree seems to work well with lightmapping).

**Comment by VReaperV** [comment]
> Or alternatively try it with lightmapping instead (the other tree seems to work well with lightmapping).

That'd probably be the easiest solution assuming it works, just a shader change.

**Comment by VReaperV** [comment]
So I was able to get this result after tweaking the shader:
![Image](https://github.com/user-attachments/assets/a352f38c-8e9c-4a17-9a7b-8dbe7aa3fe92)

However, I cannot get that result anymore. The engine sets the lighting mode for this tree to vertex because the lightmap for it is missing, which ends up looking the same as before. It's likely that the result I got above was due to some GLSL shader cache issue.

**Comment by illwieckz** [comment]
That screenshot looks like the tree is fullbright, so yes the bug can't be seen.

What I noticed is that the bug appears when we use light bounce at map compilation, but not when we don't use light bounce, so that looks a bug in the light bounce code in `q3map2`.

The good news is that I don't reproduce the bug when compiling the map with the `-sRGB` option, even with light bouncing.

**Comment by VReaperV** [comment]
> That screenshot looks like the tree is fullbright, so yes the bug can't be seen.
> 
> What I noticed is that the bug appears when we use light bounce at map compilation, but not when we don't use light bounce, so that looks like a bug in the light bounce code in `q3map2`.
> 
> The good news is that I don't reproduce the bug when compiling the map with the `-sRGB` option, even with light bouncing.

I've noticed that the shader also has `surfaceparm nolightmap`. Does that impact q3map2 lighting at all?

**Comment by VReaperV** [comment]
Or is it the other way around: q3map2 couldn't light it, and so wrote that surfaceparm?

**Comment by illwieckz** [comment]
As far as I know q3map2 doesn't write a surfaceparm that is not in a shader.

**Comment by VReaperV** [comment]
Then maybe that's why it does not create a lightmap for it?

**Comment by illwieckz** [comment]
It is probably done intentionally to use the lightgrid instead and save on generating a lightmap for all the surfaces. Now, it makes me doubt about what we do in such situation.

**Comment by illwieckz** [comment]
Well the topic is about vertex lighting, so that's likely done on purpose to use vertex lighting instead of lightmapping.

**Comment by VReaperV** [comment]
> It is probably done intentionally to use the lightgrid instead and save on generating a lightmap for all the surfaces. Now, it makes me doubt about what we do in such situation.

That's certainly not an issue now.
